Verify CHIP_ID response

mutatrum opened the pull request:
Fixes #740

Also reduces boot of BM1368 by 4 seconds. Read timeout can probably be a lot shorter for all models, but I don't have a multi-chip model to see what the actual read time is.

mutatrum commented on components/asic/bm1366.c:
```diff
@@ -244,12 +244,22 @@ static uint8_t _send_init(uint64_t frequency, uint16_t asic_count)
-    }
+
+        uint8_t checksum = crc5(asic_response_buffer, 10);
+        if (asic_response_buffer[10] != checksum) {
```
It works on the Supra, but maybe it's a 1/256 chance that was an very unlucky collision. Comparison to 0 actually didn't work.

mutatrum replied:
Ok, I didn't read properly and I was unlucky:

CRC-5 over 10, 11 is checksum:
```
aa 55 13 97 00 00 00 00 00 00 -> 11
aa 55 13 66 00 00 00 00 00 00 -> 07
aa 55 13 68 00 00 00 00 00 00 -> 0f
aa 55 13 70 00 00 00 00 00 00 -> 01
```

1368 actually gave the correct value.

CRC-5 over 9 bytes, skipping pre-amble (values taken from #740):
```
13 97 00 00 00 00 00 00 ?? -> ??
13 66 00 00 00 00 00 00 05 -> 00
13 68 00 00 00 00 00 00 0f -> 00
13 70 00 00 00 00 00 00 10 -> 00
```

I'll adjust the code.

Georges760 commented on components/asic/bm1397.c:
```diff
@@ -235,12 +235,34 @@ static uint8_t _send_init(uint64_t frequency, uint16_t asic_count)
+            continue;
+        }
+
+        if (memcmp(asic_response_buffer, "\xaa\x55\x13\x97\x00\x00", 6) != 0) {
```
on BM1397, CORE_NUM should be 0x18.

```
if (memcmp(asic_response_buffer, "\xaa\x55\x13\x97\x18\x00", 6) != 0) {
```

Georges760 reviewed (approved):
LGTM

skot commented:
For the purposes of esp-miner we should just verify the CHIP_ID (0x1366, for example) and maybe the checksum. If there is some oddball ASIC variant out there that sends something different, someone will build 1000 bitaxe with it and be very upset esp-miner doesn't work.

mutatrum commented:
I got tired of the duplicated code in the bm####.c files, so did some refactoring and clean-up:

- Moved `count_asic_chips` and `receive_work` functions to `common.c`
- Moved `asic_response_buffer` to local scope
- Unified preamble and crc check on serial rx
- Fixed typo in proccess_work
- Moved `CRC5_MASK` define to proper location

mutatrum commented on components/asic/common.c:
```diff
@@ -24,4 +34,95 @@ int _largest_power_of_two(int num)
+    }
+
+    if (received == 0) {
+        ESP_LOGE(TAG, "UART timeout in serial RX");
```
`asic_timeout_counter` in the old receive_work never increased to 2, as it is local scope. Now it occasionally logs an error, so I don't know if we should increase the timeout here, or just ignore a timeout (e.g. no logging).
